jitrebalance/jitrebalance.py

```diff
@@ -0,0 +1,184 @@
+    route += [{
+        'id': plugin.node_id,
+        'channel': scid,
+        'direction': chan['direction'],
+        'msatoshi': amt,
+        'amount_msat': '{}msat'.format(amt),
+        'delay': 9
+    }]
+
+    return route
+
+
+def try_rebalance(scid, chan, amt, peer, request):
+    # Exclude the channel we are trying to rebalance when searching for a
+    # path. We will manually append it to the route and bump the other
+    # parameters so it can be used afterwards
+    exclusions = [
+        "{scid}/{direction}".format(scid=scid, direction=chan['direction'])
+    ]
+
+    # Try up to 5 times to rebalance that last leg.
+    for i in range(0, 5):
+        route = get_circular_route(scid, chan, amt, peer, exclusions, request)
+
+        # We're about to initiate a rebalancing, we'd better remember how we can
+        # settle it once we see it back here.
+        payment_key = secrets.token_bytes(32)
+        payment_hash = hashlib.sha256(payment_key).hexdigest()
+        plugin.rebalances[payment_hash] = {
+            "payment_key": binascii.hexlify(payment_key).decode('ASCII'),
+            "payment_hash": payment_hash,
+            "request": request,
+        }
+
+        # After all this work we're finally in a position to judge whether a
+        # rebalancing is worth it at all. The rebalancing is considered worth it
+        # if the fees we're about to pay are less than or equal to the fees we get
+        # out of forwarding the payment.
+        plugin.log("Sending rebalance request using payment_hash={}, route={}".format(
+            payment_hash, route
+        ))
+        try:
+            plugin.rpc.sendpay(route, payment_hash)
+            # If the attempt is successful, we acknowledged it on the
+            # receiving end (a couple of line above), so we leave it dangling
+            # here.
+            plugin.rpc.waitsendpay(payment_hash)
+            return
+        except RpcError as e:
+            error = e.error['data']
+            erring_channel = error['erring_channel']
+            exclusions.append(erring_channel)
+            plugin.log("Excluding {} due to a failed attempt".format(erring_channel))
+
+    request.set_result({"result": "continue"})
+
+
+@plugin.async_hook("htlc_accepted")
+def on_htlc_accepted(htlc, onion, plugin, request, **kwargs):
+    plugin.log("Got an incoming HTLC htlc={}, onion={}".format(htlc, onion))
+
+    # The HTLC might be a rebalance we ourselves initiated, better check
+    # against the list of pending ones.
+    rebalance = plugin.rebalances.get(htlc['payment_hash'], None)
+    if rebalance is not None:
+        # Settle the rebalance, before settling the request that initiated the
+        # rebalance.
+        request.set_result({
+            "result": "resolve",
+            "payment_key": rebalance['payment_key']
+        })
+
+        # Now wait for it to settle correctly
+        # TODO Maybe be a bit smarter than having a fixed timeout here?
```

Comment:
Maybe we could generate an invoice when initiating the rebalance, then listen for `invoice_payment` ? Or maybe it would be even  slower..

Reply:
Hm, we could also just wait using `waitsendpay` or listen to the `sendpay_success` notification. I was just lazy :wink: